### PR TITLE
HPCC-19727 Add support for scapps style namespaces to dynamicESDL

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by afishbeck** [comment]
@jgao @rpastrana please review

**Review by rpastrana — COMMENTED** [review]
@afishbeck looks good, some comments.

File: esp/services/esdl_svc_engine/esdl_binding.cpp [hunk 1]
```diff
@@ -1763,15 +1773,24 @@ StringBuffer & EsdlBindingImpl::generateNamespace(IEspContext &context, CHttpReq
     if (!m_staticNamespace.isEmpty())
         return ns.set(m_staticNamespace);
 
-    if (m_pESDLService->m_serviceNameSpaceBase.length()>0)
-    {
-        ns.appendf("%s%c%s", m_pESDLService->m_serviceNameSpaceBase.str(), m_pESDLService->m_usesURLNameSpace ? '/' : ':', serv);
+    if (m_pESDLService->m_serviceNameSpaceBase.length()==0)
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_binding.cpp]: StringAttr has "isEmpty()" StringBuffer didn't used to, but I'll look.

File: esp/services/esdl_svc_engine/esdl_binding.cpp [hunk 2]
```diff
@@ -1763,15 +1773,24 @@ StringBuffer & EsdlBindingImpl::generateNamespace(IEspContext &context, CHttpReq
...
+            throw MakeStringExceptionDirect(ERR_ESDL_BINDING_INTERNERR, "Could not generate namespace, ensure ESDL definition is correctly configured.");
+        ns.set(SCAPPS_NAMESPACE_BASE).append(esdlServiceName);
+        return ns;
     }
-    else
-        throw MakeStringExceptionDirect(-1, "Could not generate namespace, ensure namespace base is correctly configured.");
+
+    ns.appendf("%s%c%s", m_pESDLService->m_serviceNameSpaceBase.str(), m_pESDLService->m_usesURLNameSpace ? '/' : ':', serv);
+
+    if (method && strlen(method) > 0)
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_binding.cpp]: Hmmm. this code was just moved, but I'll change it.

File: esp/services/esdl_svc_engine/esdl_binding.cpp [hunk 3]
```diff
@@ -162,8 +162,13 @@ void EsdlServiceImpl::init(const IPropertyTree *cfg,
     m_bGenerateLocalTrxId = true;
     m_custTrxCompileFail = false;
 
-    VStringBuffer xpath("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
-    IPropertyTree * srvcfg = cfg->queryPropTree(xpath.str());
+    VStringBuffer xpath("Software/EspProcess[@name=\"%s\"]", process);
+    IPropertyTree * espcfg = cfg->queryPropTree(xpath);
+    if (!espcfg)
+        throw MakeStringException(-1, "Could not access ESP process configuration: esp process '%s' service name '%s'", process, service);
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_binding.cpp]: right, cut and paste mostly.  I'll add a code.

**Comment by afishbeck** [comment]
@gaojx

**Comment by afishbeck** [comment]
@rpastrana updated with your suggestions.

**Review by rpastrana — APPROVED** [review]
@afishbeck

**Comment by afishbeck** [comment]
@richardkchapman please merge.

**Pull request merged**